provisioning_client/adapters/hsm_client_riot.c

```diff
@@ -155,33 +136,49 @@ static const HSM_CLIENT_X509_INTERFACE x509_interface =
     hsm_client_riot_get_common_name
 };
 
-static int generate_root_ca_info(HSM_CLIENT_X509_INFO* riot_info, RIOT_ECC_SIGNATURE* tbs_sig)
+// Free the mbedtls_mpi members of the signature
+static void ecc_signature_destroy(RIOT_ECC_SIGNATURE* tbs_sig)
+{
+    mbedtls_mpi_free(&tbs_sig->r);
+    mbedtls_mpi_free(&tbs_sig->s);
+}
+
+static void x509_cert_free(RIOT_ECC_PUBLIC* pub, RIOT_ECC_PRIVATE* priv)
+{
+    mbedtls_ecp_point_free(pub);
+    mbedtls_mpi_free(priv);
+}
+
+static int generate_root_ca_info(HSM_CLIENT_X509_INFO* riot_info)
 {
     int result;
     uint8_t der_buffer[DER_MAX_TBS] = { 0 };
     DERBuilderContext der_ctx = { 0 };
     DERBuilderContext der_pri_ctx = { 0 };
     RIOT_STATUS status;
-
-    memcpy(&riot_info->ca_root_pub, eccRootPubBytes, sizeof(ecc_publickey));
-    memcpy(&riot_info->ca_root_priv, eccRootPrivBytes, sizeof(ecc_privatekey));
+    RIOT_ECC_SIGNATURE tbs_sig = { 0 };
 
     // Generating "root"-signed DeviceID certificate
     DERInitContext(&der_ctx, der_buffer, DER_MAX_TBS);
     DERInitContext(&der_pri_ctx, der_buffer, DER_MAX_TBS);
-
-    if (X509GetDeviceCertTBS(&der_ctx, &X509_ROOT_TBS_DATA, &riot_info->ca_root_pub) != 0)
+    if ((status = RiotCrypt_DeriveEccKey(&riot_info->ca_root_pub, &riot_info->ca_root_priv,
+        g_digest, RIOT_DIGEST_LENGTH, (const uint8_t*)RIOT_LABEL_ALIAS, lblSize(RIOT_LABEL_ALIAS))) != RIOT_SUCCESS)
+    {
+        LogError("Failure: RiotCrypt_DeriveEccKey returned invalid status %d.", status);
+        result = MU_FAILURE;
+    }
+    else if (X509GetDeviceCertTBS(&der_ctx, &X509_ROOT_TBS_DATA, &riot_info->ca_root_pub, (uint8_t*)&riot_info->ca_root_pub, sizeof(riot_info->ca_root_pub)) != 0)
```

Comment:
This should be a call to X509GetRootCertTBS not X509GetDeviceCertTBS.

Reply:
Fixed